// @dicy/core/lib/Rules/ParseXindyLog.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _set = require('babel-runtime/core-js/set');

var _set2 = _interopRequireDefault(_set);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _Rule = require('../Rule');

var _Rule2 = _interopRequireDefault(_Rule);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ParsedXindyLog extends _Rule2.default {

  /**
   * Parse the xindy log.
   * @return {Promise<boolean>}  Status of rule evaluation
   */
  parse() {
    var _this = this;

    return (0, _asyncToGenerator3.default)(function* () {
      // Get the output file
      const output = yield _this.getResolvedOutput('$FILEPATH_0-ParsedXindyLog');
      if (!output) return false;

      const parsedLog = {
        messages: [],
        inputs: [],
        outputs: [],
        calls: []
      };
      const name = 'xindy';
      let filePath;

      yield _this.firstParameter.parse([{
        // Error/Warning messages with file references
        names: ['severity', 'text', 'file'],
        patterns: [/^(ERROR|WARNING): (.*? in:?)$/i, /(.*)/],
        evaluate: function (reference, groups) {
          parsedLog.messages.push({
            name,
            severity: groups.severity.toLowerCase(),
            text: `${groups.text} ${groups.file}`,
            source: { file: _path2.default.normalize(groups.file) },
            log: reference
          });
        }
      }, {
        // Error/Warning messages
        names: ['severity', 'text'],
        patterns: [/^(ERROR|WARNING): (.*)$/i],
        evaluate: function (reference, groups) {
          parsedLog.messages.push({
            name,
            severity: groups.severity.toLowerCase(),
            text: groups.text,
            source: { file: filePath },
            log: reference
          });
        }
      }, {
        // Module loadings
        names: ['text'],
        patterns: [/^(Loading module .*|Finished loading module .*)$/i],
        evaluate: function (reference, groups) {
          parsedLog.messages.push({
            name,
            severity: 'info',
            text: groups.text,
            source: { file: filePath },
            log: reference
          });
        }
      }, {
        // Input files
        names: ['file'],
        patterns: [/^Reading raw-index (.*?)[.]{3}$/i],
        evaluate: function (reference, groups) {
          filePath = _path2.default.normalize(groups.file);
          parsedLog.inputs.push(_path2.default.normalize(groups.file));
          parsedLog.messages.push({
            name,
            severity: 'info',
            text: groups._,
            source: { file: filePath },
            log: reference
          });
        }
      }, {
        // Output files
        names: ['file'],
        patterns: [/^(?:Markup written into file|Opening logfile) (.*?)[. ]$/i],
        evaluate: function (reference, groups) {
          parsedLog.outputs.push(_path2.default.normalize(groups.file));
          parsedLog.messages.push({
            name,
            severity: 'info',
            text: groups._,
            source: { file: filePath },
            log: reference
          });
        }
      }]);

      output.value = parsedLog;

      return true;
    })();
  }
}
exports.default = ParsedXindyLog;
ParsedXindyLog.parameterTypes = [new _set2.default(['XindyLog'])];
ParsedXindyLog.commands = new _set2.default(['build', 'log']);
ParsedXindyLog.description = 'Parses the logs produced by xindy and texindy.';
ParsedXindyLog.defaultActions = ['parse'];
