// @dicy/core/lib/Rules/ParseBibToGlsLog.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _set = require('babel-runtime/core-js/set');

var _set2 = _interopRequireDefault(_set);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _Rule = require('../Rule');

var _Rule2 = _interopRequireDefault(_Rule);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ParseBibTeXLog extends _Rule2.default {

  parse() {
    var _this = this;

    return (0, _asyncToGenerator3.default)(function* () {
      const output = yield _this.getResolvedOutput('$FILEPATH_0-ParsedBibToGlsLog');
      if (!output) return false;

      const parsedLog = {
        messages: [],
        inputs: [],
        outputs: [],
        calls: []
      };

      yield _this.firstParameter.parse([{
        names: ['output'],
        patterns: [/^Writing (.*)$/],
        evaluate: function (reference, groups) {
          parsedLog.outputs.push(_this.normalizePath(groups.output));
        }
      }, {
        names: ['input'],
        patterns: [/^Reading (.*)$/],
        evaluate: function (reference, groups) {
          parsedLog.inputs.push(_this.normalizePath(groups.input));
        }
      }]);

      output.value = parsedLog;

      return true;
    })();
  }
}
exports.default = ParseBibTeXLog;
ParseBibTeXLog.parameterTypes = [new _set2.default(['BibToGlsLog'])];
ParseBibTeXLog.commands = new _set2.default(['build', 'log']);
ParseBibTeXLog.defaultActions = ['parse'];
ParseBibTeXLog.description = 'Parses any bib2gls produced logs.';
