// @dicy/core/lib/Rules/ParseLaTeXAuxilary.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _set = require('babel-runtime/core-js/set');

var _set2 = _interopRequireDefault(_set);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _Rule = require('../Rule');

var _Rule2 = _interopRequireDefault(_Rule);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ParseLaTeXAuxilary extends _Rule2.default {

  parse() {
    var _this = this;

    return (0, _asyncToGenerator3.default)(function* () {
      const output = yield _this.getResolvedOutput('$FILEPATH_0-ParsedLaTeXAuxilary');
      if (!output) return false;

      const results = {
        commands: []

        // We just need a list of the commands that in the aux file so we know
        // when to run BibTeX and Bib2Gls.
      };yield _this.firstParameter.parse([{
        names: ['command'],
        patterns: [/^\\([A-Za-z@]+)/],
        evaluate: function (reference, groups) {
          results.commands.push(groups.command);
        }
      }]);

      output.value = results;

      return true;
    })();
  }
}
exports.default = ParseLaTeXAuxilary;
ParseLaTeXAuxilary.parameterTypes = [new _set2.default(['LaTeXAuxilary'])];
ParseLaTeXAuxilary.defaultActions = ['parse'];
ParseLaTeXAuxilary.description = 'Parses the aux files produced by all variants of LaTeX.';
