// @dicy/core/lib/Rules/PdfToPs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _set = require('babel-runtime/core-js/set');

var _set2 = _interopRequireDefault(_set);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _State = require('../State');

var _State2 = _interopRequireDefault(_State);

var _File = require('../File');

var _File2 = _interopRequireDefault(_File);

var _Rule = require('../Rule');

var _Rule2 = _interopRequireDefault(_Rule);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class PdfToPs extends _Rule2.default {

  static isApplicable(state, command, phase, options, parameters = []) {
    return (0, _asyncToGenerator3.default)(function* () {
      // Only apply if output format is ps
      return options.outputFormat === 'ps';
    })();
  }

  initialize() {
    var _this = this;

    return (0, _asyncToGenerator3.default)(function* () {
      // Zap the previous target since we are building a pdf
      yield _this.replaceResolvedTarget('$FILEPATH_0', '$DIR_0/$NAME_0.ps');
    })();
  }

  constructCommand() {
    return {
      args: ['pdf2ps', '{{$FILEPATH_0}}', '{{$DIR_0/$NAME_0.ps}}'],
      cd: '$ROOTDIR',
      severity: 'error',
      outputs: ['$DIR_0/$NAME_0.ps']
    };
  }
}
exports.default = PdfToPs;
PdfToPs.parameterTypes = [new _set2.default(['PortableDocumentFormat'])];
PdfToPs.description = 'Converts PDF to PS using pdf2ps. Enabled by the `pdfProducer` option.';
